### ISSUE TITLE
Offline Usage of OSPOS "Host name lookup failure"

### ISSUE THREAD

**Issue by Modding** [jekkos/opensourcepos#965]
Hi i was trying to use the OSPOS on my offline POS System, and it awlays says:

A PHP Error was encountered
Servetity: Warning

Message: socket_connect(): Host loockup failed [-10002]: Host name lookup failure
Filename: Adapter/Socket.php

Line Number: 32

Backtrace:
[...]

And so on ...

If i connect the POS System to a Network with Internet it works.

Do i have to change something in the Config or is it a Bug?

**Comment by onyema3** [comment]
No such errors with the latest masters. Give that a try

**Comment by Modding** [comment]
it is a master checked out 5 minutes ago

**Comment by Modding** [comment]
Here is the full message:

`A PHP Error was encountered

Severity: Warning

Message: socket_connect(): Host lookup failed [-10002]: Host name lookup failure

Filename: Adapter/Socket.php

Line Number: 32

Backtrace:

File: /var/www/html/vendor/ins0/google-measurement-php-client/src/Racecore/GATracking/Client/Adapter/Socket.php
Line: 32
Function: socket_connect

File: /var/www/html/vendor/ins0/google-measurement-php-client/src/Racecore/GATracking/Client/Adapter/Socket.php
Line: 114
Function: createConnection

File: /var/www/html/vendor/ins0/google-measurement-php-client/src/Racecore/GATracking/GATracking.php
Line: 339
Function: send

File: /var/www/html/vendor/ins0/google-measurement-php-client/src/Racecore/GATracking/GATracking.php
Line: 370
Function: callEndpoint

File: /var/www/html/application/libraries/Tracking_lib.php
Line: 87
Function: sendTracking

File: /var/www/html/application/controllers/Login.php
Line: 31
Function: track_page

File: /var/www/html/public/index.php
Line: 315
Function: require_once
`

**Comment by onyema3** [comment]
Do you have php-sockets installed and enabled?

**Comment by Modding** [comment]
Of course ... if it wouldnt be enabled it wouldnt working in "online" modus.

**Comment by Modding** [comment]
Is there a Workaround? I need the POS System on Saturday in a venue i havent got Internet.

**Comment by daN4cat** [comment]
So php-sockets is enabled but with no Internet connection you get the error anyway.
I need to check this because I fixed in the past but it seems not completely right.

**Comment by onyema3** [comment]
Mine is working just fine offline. Did you change anything??

**Comment by daN4cat** [comment]
He's not the first to report this. So I need to re-check...

**Comment by jekkos** [comment]
Indeed I'm facing the same issue even after disabling the db option.

**Comment by daN4cat** [comment]
@jekkos since you can reproduce this (I cannot as I don't have the right setup), can you please check a change for me?

Move the following in Socket.php

```
        if ($this->getOption('async')) {
            socket_set_nonblock($connection);
        }
```

to be before `if (!socket_connect` line and test again.

**Comment by daN4cat** [comment]
@Modding you can try the same change too

**Comment by daN4cat** [comment]
Searching around it seems more an issue with the setup: http://serverfault.com/a/16021

Please ignore what I suggested earlier and check what that comment suggests.

**Comment by jekkos** [comment]
So it's just a php glitch in this case? Restart would then solve it.? I found the error in the docker container..

**Comment by daN4cat** [comment]
Maybe I should ring-fence the Socket createConnection internals with a call to `gethostbyaddr ( string $ip_address )` and see if in that way we get a FALSE without throwing an error.
If FALSE then return without going any further.

But I would need some help to test because I cannot reproduce this.

Do you have time to give a try adding this as first lines in that function?

```
// IP address is www.google.com
if(gethostbyaddr ('216.58.204.68') === FALSE)
{
    return FALSE;
}
```

**Comment by Modding** [comment]
So im in office right now ... the php glitch is not the Problem.
https://www.youtube.com/watch?v=FTR01vqBEHI

I will try the suggestion that daN4cat in 5 minutes

**Comment by daN4cat** [comment]
OK, let me know how it goes. So in Socket.php just after:

```
    private function createConnection($endpoint)
    {
```

add the code I suggested above.

**Comment by Modding** [comment]
So i put this in the Socket.php ... Same errors

**Comment by daN4cat** [comment]
Can you dump the error please?

**Comment by daN4cat** [comment]
@Modding something else to try.

In `public/index.php`, can you please comment this line:

`//	define('ENVIRONMENT', isset($_SERVER['CI_ENV']) ? $_SERVER['CI_ENV'] : 'development');`

and write this one below:

`	define('ENVIRONMENT', isset($_SERVER['CI_ENV']) ? $_SERVER['CI_ENV'] : 'production');`

See if you still get the error notification on the screen or not.

**Comment by daN4cat** [comment]
I probably need to do that change in master as I remember I did some time ago but it seems it went lost/changed?

**Comment by Modding** [comment]
That works! PERFECT!

**Comment by Modding** [comment]
https://www.youtube.com/watch?v=ZJVeAUpzv5w

**Comment by daN4cat** [comment]
Great, I'll submit that change to master because it should be production as default not development anyway.

The error is not shown to the user and in any case it's not relevant.

BTW, I see you are running OSPOS on Raspberry PI, how do you find it? is it working fine? It seems responsive enough.
What's your setup in detail? We have one documented in the wiki but there was an alternative that seems good but not sure if something dare to test.

**Comment by Modding** [comment]
I am Running OSPOS Native on an Raspi3 in PI-Top Ceed case with Raspberrian/Debian on it.
It works well and its fully functional.
Ive written a Script, that automaticly opens Chromium in Kiosk mode if you Boot up. The Apache and MySQL works perfect as expected and brings no errors and so on. 
The Barcode Scanner comes as HUID and works like a Keyboard out of the Box.

Ive alterd some views like the Reciept Default view to work with the Printer and my Cash Drawyer. 
I used the linux Command lp and the default Cups generic Text Only Printer with some changes to open The Cash Drawyer ...
I also put in a line of code that Pics one Cupon code out of a stash of cupon codes if anyone buys a CD they get the Code for downloading our CD on theyr digital devices ...

I have never used OSPOS on this Setup in an Production enviroment. On Saturday we will use the System for the First time at our Merch booth at a Concert in Munich. 

I also installed a Wireless EC Terminal to get CC and EC payments. But this is not managed by the POS System (yet) it is an Android App with a little Bluetooth Terminal that costs me 1% of the Sum with EC ... This system is called Sumup: http://fbuy.me/eykDQ

This is my Actual Setup, that i will test on Saturday at our Concert.

**Comment by daN4cat** [comment]
Sounds good, good luck for your Saturday sale. Let me know if all worked well.
Good to know about the Raspi3 setup.
Maybe you can write something in the wiki about the printer setup and what you did to deal with the cash drawer. Would be nice to share a bit of info about that.

**Comment by daN4cat** [comment]
@Modding I forgot to recommend disabling the xss clean functionality given it's a standalone and not connected installation. That will speed up the performance or reports and similar type of functionality.
That's a security measure necessary only if you run OSPOS from the Internet.

That configuration can be disabled in config/config.php like this:

`$config['ospos_xss_clean'] = FALSE;`

**Comment by dugganjl** [comment]
I also encountered this error whether connected to the internet or not, and found that it was caused by an improperly configured /etc/network/interfaces file. I had recently modified the file to set a static IP for my ethernet port, but had mistakenly omitted the 'auto enp2s0' in the following config:

```
auto enp2s0
iface enp2s0 inet static
address 192.168.1.74
gateway 192.168.1.254
netmask 255.255.255.0
dns-nameservers 8.8.8.8 8.8.4.4
```

Upon including the 'auto enp2s0' (which is my adapter's name), the problem was solved. I did not first try any other advice in this thread.
